engines/docker/resultset.go

```diff
@@ -31,115 +41,189 @@ func (r *resultSet) Success() bool {
-		hdr, err := reader.Next()
-		if err != nil {
-			break
-		}
-		if hdr.Typeflag == tar.TypeDir {
-			continue
-		}
-		fname := strings.TrimPrefix(hdr.Name, strip)
-		m := monitor.WithTag("filename", fname)
 
-		tempfile, err := r.storage.NewFile()
-		if err != nil {
-			m.ReportError(err, "could not create temporary file")
-			continue
-		}
+	// Read tar stream while we download from docker
+	stream, streamWriter := io.Pipe()
 
-		if _, err = io.Copy(tempfile, reader); err != nil {
-			m.ReportError(err, "could not copy file")
-			continue
-		}
+	ctx, cancel := context.WithCancel(r.context)
+	notFound := false
+	var derr, rerr error
+	var interrupted atomics.Once
+	util.Parallel(func() {
+		defer cancel()
+		defer stream.Close() // always close the reader, so writers abort instead of hanging
+		// Create tar reader
+		reader := tar.NewReader(stream)
+		var wg atomics.WaitGroup
+		for !interrupted.IsDone() {
+			// Get first entry from reader
+			var hdr *tar.Header
+			hdr, rerr = reader.Next()
+			if rerr == io.EOF {
+				rerr = nil // EOF is not an error
+				return
+			}
+			if rerr != nil {
+				rerr = errors.Wrap(rerr, "failed to read tar-stream from docker")
+				return
+			}
+			if !ioext.IsPlainFileInfo(hdr.FileInfo()) {
+				continue // skip entries that aren't files
+			}
+			// Sanity check on the file size mostly in case someone tries to export
+			// a sparse file that contains a lot of zeros.
+			if hdr.Size > maxExtractedFileSizeAllowed*1024*1024*1024 {
+				rerr = fmt.Errorf(
+					"attempted to extract file: '%s' at size: %d larger than allowed %d GiB",
+					path, hdr.Size, maxExtractedFileSizeAllowed,
+				)
+				r.context.LogError(fmt.Sprintf(
+					"A plugin attempted to read the file: '%s' which has a size: %d greater than %d GiB, "+
+						"which is the maximum allowed file size hardcoded into the worker for sanity",
+					path, hdr.Size, maxExtractedFileSizeAllowed,
+				))
+				return
+			}
+			debug("ExtractFolder(%s) found file: '%s' of size: %d bytes", path, hdr.Name, hdr.Size)
+
+			// Create temporary file that we extract this to
+			var tmpfile runtime.TemporaryFile
+			tmpfile, rerr = r.storage.NewFile()
+			if rerr != nil {
+				rerr = errors.Wrap(rerr, "failed to create temporary file for extracting files from docker")
+				return
+			}
 
-		_, err = tempfile.Seek(0, 0)
-		if err != nil {
-			m.ReportWarning(err, "could not seek to start of file")
-			continue
+			// Copy data to tmpfile
+			if _, rerr = io.Copy(tmpfile, reader); rerr != nil {
+				tmpfile.Close()
+				rerr = errors.Wrap(rerr, "failed to create temp-file using tar-stream from docker")
+				return
+			}
+
+			// Seek start of temp file
+			if _, rerr = tmpfile.Seek(0, io.SeekStart); rerr != nil {
+				tmpfile.Close()
+				rerr = errors.Wrap(rerr, "failed to seek to start of temporary file, after reading from docker")
+				return
+			}
+
+			// Invoke handler concurrently
+			wg.WaitForLessThan(5) // No more than 5 concurrent calls to handler
```

Comment:
I see this is pointless, the Go scheduler already leverages concurrency pretty well.

Reply:
yeah, but we're creating temporary files... So maybe only having 5 of those around at the same time is a good sanity check..

Maybe it should be higher, but some limit to concurrency is probably a good idea...

Reply:
We then should ignore it for now, just setting an arbitrary number in concurrency doesn't fix the problem.

Reply:
Other engines always call the handler sequentially. So I think 5 is good.

But I'll make it a constant so it's easy to identify. We have these sanity limits many places, some we'll never need to tweak others will need tweaks in the future.